[PM-7330] Create SetPasswordComponent (email verification)

rr-bw opened the pull request:
## 🎟️ Tracking

- [PM-7330](https://bitwarden.atlassian.net/browse/PM-7330) - Create new SetPasswordComponent
- [PM-5133](https://bitwarden.atlassian.net/browse/PM-5133) - [Web] JIT-provisioned SSO user should use new SetPasswordComponent
- [PM-7415](https://bitwarden.atlassian.net/browse/PM-7415) - [Browser] JIT-provisioned SSO user should use new Set Password component
- [PM-7416](https://bitwarden.atlassian.net/browse/PM-7416) - [Desktop] JIT-provisioned SSO user should use new Set Password component

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->

## 📔 Objective

Creates a `SetPasswordJitComponent` and `DefaultSetPasswordJitService` in Libs. Web and Desktop each have some client-specific code to run after the password is set, so each of these clients has a service that extends the default service and overrides where necessary.

Much of the logic for the SetPasswordComponent is copied (with some refactoring) from our [existing SetPasswordComponent](https://github.com/bitwarden/clients/blob/main/libs/angular/src/auth/components/set-password.component.ts) and then modified to work with our new `InputPasswordComponent`.

## 📸 Screencasts

### Web

https://github.com/bitwarden/clients/assets/102181210/17b0ba40-935b-4f64-8ebb-154bf2d0412e

### Desktop

https://github.com/bitwarden/clients/assets/102181210/535f602f-8417-45c8-9fc9-4cdf49479ff6

### Browser

Note: the Sync issue at the end of this video has been resolved. It was just an issue with my local env, but is now fixed.

https://github.com/bitwarden/clients/assets/102181210/c4c210c8-fc4d-4044-8548-867061167fa8

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes


[PM-7330]: https://bitwarden.atlassian.net/browse/PM-7330?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PM-5133]: https://bitwarden.atlassian.net/browse/PM-5133?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[PM-7415]: https://bitwarden.atlassian.net/browse/PM-7415?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PM-7416]: https://bitwarden.atlassian.net/browse/PM-7416?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jlf0dev commented on libs/auth/src/angular/set-password-jit/set-password-jit.component.ts:
```diff
@@ -0,0 +1,121 @@
+      try {
+        this.orgSsoIdentifier = qParams.identifier;
+
+        const autoEnrollStatus = await this.organizationApiService.getAutoEnrollStatus(
+          this.orgSsoIdentifier,
+        );
+        this.orgId = autoEnrollStatus.id;
+        this.resetPasswordAutoEnroll = autoEnrollStatus.resetPasswordEnabled;
+        this.masterPasswordPolicyOptions =
+          await this.policyApiService.getMasterPasswordPolicyOptsForOrgUser(autoEnrollStatus.id);
```
I think the auto enroll logic should belong in the service. I would pass the orgId to the service if present and resolve the auto enroll based on the existence.

jlf0dev commented on libs/auth/src/angular/set-password-jit/default-set-password-jit.service.ts:
```diff
@@ -0,0 +1,147 @@
+    );
+  }
+
+  async runClientSpecificLogicAfterSetPasswordSuccess(): Promise<void> | null {
+    // override in client-specific service
+    return null;
+  }
```
Even though I like the fact that this doesn't overcomplicate things, the service shouldn't be aware that there is client specific logic needed to be run. This also requires the caller to know to call this after the `setPassword`.

Instead, we should override the `setPassword` method in each client specific service, call the original method through `super`. Then add the client specific logic after.

jlf0dev commented on libs/auth/src/angular/set-password-jit/set-password-jit.component.ts:
```diff
@@ -0,0 +1,121 @@
+        this.userId,
+      );
+      await this.setPasswordJitService.runClientSpecificLogicAfterSetPasswordSuccess();
+    } catch (e) {
+      this.validationService.showError(e);
+      return;
+    }
```
do we need to set `this.submitting = false` here?

jlf0dev commented on libs/auth/src/angular/set-password-jit/set-password-jit.service.abstraction.ts:
```diff
@@ -0,0 +1,25 @@
+import { PasswordInputResult } from "../input-password/password-input-result";
+
+export abstract class SetPasswordJitService {
+  /**
+   * Uses the output from InputPasswordComponent to set a password
+   * for a JIT provisioned user.
+   */
```
`InputPasswordComponent` shouldn't be referenced from the service, it couples them too much for my liking. Please provide `@param` definitions and `@throws` as well.

jlf0dev commented on libs/auth/src/angular/set-password-jit/set-password-jit.service.abstraction.ts:
```diff
@@ -0,0 +1,25 @@
+   * for a JIT provisioned user.
+   */
+  setPassword: (
+    passwordInputResult: PasswordInputResult,
```
I'd rather have a model created specifically for this service than reusing the `PasswordInputResult` here. Just to decouple things.

jlf0dev commented on libs/auth/src/angular/set-password-jit/default-set-password-jit.service.ts:
```diff
@@ -0,0 +1,147 @@
+    protected userDecryptionOptionsService: InternalUserDecryptionOptionsServiceAbstraction,
+  ) {}
+
+  async setPassword(
+    passwordInputResult: PasswordInputResult,
+    orgSsoIdentifier: string,
+    orgId: string,
+    resetPasswordAutoEnroll: boolean,
+    userId: UserId,
+  ): Promise<void> {
```
I'd like to see some error handling in this service. We need to check input parameters for nullish values, as well as the result of external services we call while assembling request (within reason of course).

jlf0dev reviewed (changes requested):
Looking good! Left some changes I'd like to see before approving, but excited about this one.

jlf0dev commented:
I also just noticed that the success toasts weren't showing in the videos, are those from the previous iteration of this PR?

Also going to tag @danielleflinn to see if she'd like to do a design review on this one, the MP policy requirements on the browser look a little...tight.

rr-bw commented:
> I also just noticed that the success toasts weren't showing in the videos, are those from the previous iteration of this PR?

Yes, those were previous videos. Here's an updated one.

https://github.com/bitwarden/clients/assets/102181210/f91d4187-924b-48e5-8a60-925418d7b843

rr-bw commented:
> I do agree that both the browser and desktop MP policy callouts look a bit off though. Is there a way we can fix the spacing between the leftmost edge of the bullets and the text?

@danielleflinn Here are updated screenshots for Desktop and Browser:

<img width="443" alt="desktop-list" src="https://github.com/bitwarden/clients/assets/102181210/31b46895-3f17-4bf3-a9f1-d26fc7162c76">

![browser-list](https://github.com/bitwarden/clients/assets/102181210/4471c1f6-23eb-4927-a1b8-c07fb58f6016)

danielleflinn commented:
@rr-bw looks great! Thank you

jlf0dev commented on libs/auth/src/angular/set-password-jit/default-set-password-jit.service.ts:
```diff
@@ -0,0 +1,184 @@
+export class DefaultSetPasswordJitService implements SetPasswordJitService {
+  orgId: string;
+
+  resetPasswordAutoEnroll: boolean;
+  masterPasswordPolicyOptions: MasterPasswordPolicyOptions;
```
I think i've steered you wrong here.

I missed that the orgId is retrieved from the autoEnroll response. I want to avoid having properties on the service, so I think we should move this back to the component. Sorry for the back and forth, this one is fully on me.

The real fix here is adding something to the server that can get us the orgId. Taking shortcuts like that always end up biting us later down the line.

jlf0dev commented on libs/auth/src/angular/set-password-jit/default-set-password-jit.service.ts:
```diff
@@ -0,0 +1,184 @@
+
+    // Since this is an existing JIT provisioned user in a MP encryption org setting first password,
+    // they will not already have a user asymmetric key pair so we must create it for them.
+    const newKeyPair = await this.cryptoService.makeKeyPair(protectedUserKey[0]);
+    if (newKeyPair == null) {
+      throw new Error("newKeyPair not found. Could not set password.");
+    }
+    const keysRequest = new KeysRequest(newKeyPair[0], newKeyPair[1].encryptedString);
```
nit: this could probably be its own method

rr-bw commented:
@jlf0dev I addressed the recent feedback and also updated the PR description with a new screencast for each client. I am having one issue with Browser. As you can see in the video, setting the password works fine, but I cannot get the account to sync with the org and get access to the org. Here is the error I get:

![Screenshot 2024-07-03 at 11 38 35 AM](https://github.com/bitwarden/clients/assets/102181210/64e75ca2-07fd-40b5-a1eb-3574e0082796)

Am I potentially doing something wrong with service injection? I also tried adding the following to `main.background.ts`, but that did not help.

![Screenshot 2024-07-03 at 12 33 56 PM](https://github.com/bitwarden/clients/assets/102181210/c25e0521-37fe-4abc-8642-5334da1e6371)

danielleflinn commented:
@rr-bw I just noticed a few inconsistencies [with Figma](https://www.figma.com/design/81uGWWdQVAQfKDiR4qltoq/Email-Verification-%5BCreate-account%5D?node-id=2055-64010&t=IG5IIvVUNJEbh9G8-4). Are we able to make the title of this page "Join organization" (once we can support[ dynamic text](https://bitwarden.atlassian.net/browse/PM-9603) in the Anon layout title we can update this to "Join [Org name]" to match Figma), and can we change the CTA to "Create account"?

Also, I know in Figma we have the "Finishing joining org..." text in the content area; but what do you think about moving this to the Anon layout subtitle? 

If we are using this UI for another flow other than the Join organization we can probably keep the text the same, but if it is just for the Join Org flow, lets customize it a bit for the user.

rr-bw commented:
> I am having one issue with Browser. As you can see in the video, setting the password works fine, but I cannot get the account to sync with the org and get access to the org.

Issue fixed in a conversation with @jlf0dev

rr-bw commented:
> Are we able to make the title of this page "Join organization" (once we can support[ dynamic text](https://bitwarden.atlassian.net/browse/PM-9603) in the Anon layout title we can update this to "Join [Org name]" to match Figma), and can we change the CTA to "Create account"?
>
> Also, I know in Figma we have the "Finishing joining org..." text in the content area; but what do you think about moving this to the Anon layout subtitle?

@danielleflinn For sure, here's an updated screenshot:

![Screenshot 2024-07-12 at 11 59 51 AM](https://github.com/user-attachments/assets/2b58bdaf-82b0-4180-9628-41c11a1e0290)

danielleflinn commented:
@rr-bw Thank you this looks great! I'll make sure things are updated in Figma accordingly for QA

jlf0dev commented on libs/auth/src/angular/set-password-jit/default-set-password-jit.service.ts:
```diff
@@ -0,0 +1,170 @@
+    protected userDecryptionOptionsService: InternalUserDecryptionOptionsServiceAbstraction,
+  ) {}
+
+  async setPassword(credentials: SetPasswordCredentials): Promise<void> {
```
Another thought, it might be nice here to check the user decryption options and/or master key before doing anything to make sure that the user doesn't already have a password. Just as a nice sanity check.

jlf0dev reviewed (dismissed):
Looks good! I had a couple thoughts but nothing blocking.
